### ISSUE TITLE
feature request desktop privacybar

### ISSUE THREAD

**Issue by smartekIT** [Ylianst/MeshCentral#3850]
Hi,
most of the time when we try to unpin the desktop privacy bar during desktop remote session, it freezes on us and showing nonresponding specially on older OS and server's OS e.g. win7 and windows server family. Is there away we can set it to show for a limited time then it unpin itself through our Config.js ?
or if there is an option can be set there to give us the choice to have the privacy bar unpinned or pinned by default ?
also the privacy bar itself looks too wide measuring from left to right, can we make it smaller? I think the old privacy bar size was perfect before.

Thanks

**Comment by vitko-bg** [comment]
Right click on Connect button, you should have a choice with or with no bar connection

**Comment by smartekIT** [comment]
> Right click on Connect button, you should have a choice with or with no bar connection

I'm already aware of that, but the suggestion is to still having it showing but default unpinned, or pinned for a period of time then it unpin itself automatically while not dismissed, but instead hide on top until hover the mouse to show it again.

it's good to have it because it shows the users that device being controlled, but currently it's pinned by default to show on top with that wide area, if you open browser in the remote pc, you hardly can see the tabs. and when you decide to unpin it so you can see the browser tabs properly then it freeze and become nonresponding on win7 or windows server families e.g. 2008R2, 2012, 2016 and 2019. and some times in win10 when the pc very busy with other services loading.

**Comment by smartekIT** [comment]
to be honest other remote tools has it in a better way, smaller in size and clearer pin/unpin proper button and ability to drag left or right from your way. if you logged in to a pc and then from that remote pc start windows remote desktop connection, then you can see how hard it's to reach to the built-in windows remote bar. and if freezes on you then the only way to fix is to disconnect and reconnect again and if consent required and there is no user on the other side anymore then that's a problem.

**Comment by SomeGuru** [comment]
+1 for auto hide, but given a small logo that can be hovered over to reappear as needed, but a gentle reminder the system is under mesh viewing.... Maybe like a small site/mesh logo...

**Comment by krayon007** [comment]
Are the freezing issues happening on 32 bit or 64 bit agent? (Or is it happening on both?)

**Comment by VINISHVV** [comment]
@krayon007 - its happens on both 32 and 64bit os

**Comment by smartekIT** [comment]
> @krayon007 - its happens on both 32 and 64bit os

yes happens on both. normally the older OS or windows server family OS with a high chance to freeze and blur to white all over it and being nonresponding specially when you click on it trying to click on the pin button. and by the way, when hover the mouse over the pin icon, the mouse style never change to a hand or grip for example like any other button you would expect. To see what I mean, try hover on hyper link or any windows or web buttons, mouse style changes to let you know there is a click can happen there, but with this one nothing happen, so hard to get exact click on the pin specially with the small screens so you find your self clicking next to the pin rather than the pin multiple tries then it freezes straightaway.

once it freeze then you can do nothing at all to get it back or get it away, so the only solution then is to disconnect then reconnect again. as I mentioned up some times reconnect is not an available option.

Thanks guys

**Comment by smartekIT** [comment]
![RUScreen](https://user-images.githubusercontent.com/70715469/162285067-8fa9c38a-de01-40f6-a367-29d29fde07bc.gif)
![MCScreen](https://user-images.githubusercontent.com/70715469/162285095-95efad69-3e72-4ddc-9c93-ff7a72378c9e.gif)

**Comment by smartekIT** [comment]
click on the above to examples of GIF and watch them, then you would know exactly what I mean.

Thanks

**Comment by krayon007** [comment]
Thanks. I have a couple things in mind to try to improve it. I'm currently on vacation this week, but I'll take a look at it next week. By the very least, being ae to set autohide as a configurable default should be easy.

**Comment by smartekIT** [comment]
hi @krayon007 , any luck with the above?

**Comment by smartekIT** [comment]
@krayon007 test v1.0.11 and the privacy bar still crashing and still wide as before and no autohide by default available.
please let us know if this can be fixed any soon.

Thanks

**Comment by krayon007** [comment]
I just implemented a fix for this... I'm also working on the autohide/default functionality right now...

**Comment by smartekIT** [comment]
> I just implemented a fix for this... I'm also working on the autohide/default functionality right now...

Thanks champion.

**Comment by smartekIT** [comment]
@krayon007, autohide by default still not implemented and the bar still wide on the screen. are you planning a fix soon? Thanks

**Comment by smartekIT** [comment]
> I fixed the part about it being too wide, as I set a cap on % width of display on top of the current DPI settings. I also updated it so it shows a hand icon when you hover over the pin and the X... Working on figuring out why sometimes it stops responding.

sorry for being a headache by asking too much, but when did you fix the width and the hand icon? do you mean just now or already a while ago? because I'm on the latest version and I have none of that.

Thanks champion

**Comment by krayon007** [comment]
It was fixed a while ago, but we haven't pushed the update yet as I was working on a new test tool and documentation to better test the agent, particularly update testing.

**Comment by krayon007** [comment]
We should be pushing a an update soon, as there are a few other fixes, such as webrtc that we want to deploy as well.

**Comment by smartekIT** [comment]
Thanks for the update.

**Comment by smartekIT** [comment]
hi @krayon007 @Ylianst  , I see you had updated the agents couple of times but none of these enhancements  we talked about were included.
privacy bar still the same !
still no GUI option to reboot the remote pc into SafeMode !
did you draw back on these enhancements ?

**Comment by krayon007** [comment]
We haven't updated the agent yet as I've been trying to fix some memory leak issues. But this next version has the ability to set autohide as default, as well as setting a maximum width/ratio for the bar size

**Comment by smartekIT** [comment]
Hi @krayon007. hope you are well.
Is this been included in the latest release published couple of days ago? any hit where I can enable the AutoHide option?
mouse still showing as curser when hover on top of the clip. it never change to hand grip.

Thanks

**Comment by krayon007** [comment]
Yes, it's included. For now, I created a console command called, "taskbar". If you type that in the console tab, it'll show you how to use it. I'll have to get Ylian to modify the webui to add it as an option

**Comment by smartekIT** [comment]
thanks,
is the bar meant to stay like this, no changes to the size and the mouse behaviour. I thought you said you will recreate it for something better. scroll up here for the GIF screenshots and showed.

**Comment by smartekIT** [comment]
![image](https://user-images.githubusercontent.com/70715469/187597421-8ec6b3f9-5fda-48d8-8ab7-76545caf4123.png)

**Comment by si458** [comment]
> ![image](https://user-images.githubusercontent.com/70715469/187597421-8ec6b3f9-5fda-48d8-8ab7-76545caf4123.png)
> 

Have u updated to the latest version 1.0.76?
and also updated the end devices to the latest agent 28th August 2022? `agentupdate` in console

**Comment by krayon007** [comment]
The mouse behavior was interesting. I noticed that when in some systems the mouse only changed to a hand when I moved the mouse locally not remotely. I also saw similar behavior when you connect to an rdp session. The change for the hand icon is definitely there, I just need to figure out why in some scenarios that event doesn't trigger.

**Comment by smartekIT** [comment]
> > ![image](https://user-images.githubusercontent.com/70715469/187597421-8ec6b3f9-5fda-48d8-8ab7-76545caf4123.png)
> 
> Have u updated to the latest version 1.0.76? and also updated the end devices to the latest agent 28th August 2022? `agentupdate` in console

I'm on 1.0.76

**Comment by smartekIT** [comment]
for the agent, this is what i get when under the device console tab:
<pre>

  "openssl": "1.1.1n",
  "duktape": "v2.6.0",
  "commitDate": "2022-03-25T18:19:00.000Z",

</pre>

**Comment by smartekIT** [comment]
even if i send command "agentupdate" in the device console tab, still shows the same info after update finishes successfully

**Comment by si458** [comment]
> for the agent, this is what i get when under the device console tab:
> <pre>
> 
>   "openssl": "1.1.1n",
>   "duktape": "v2.6.0",
>   "commitDate": "2022-03-25T18:19:00.000Z",
>   "compileTime": "12:04:02, Mar 25 2022",
> 
> </pre>

That's an old version, you need to issue `agentupdate` from the device console and then check again with `info` from the device console

**Comment by smartekIT** [comment]
already done that, look here https://github.com/Ylianst/MeshCentral/issues/3850#issuecomment-1232491974

**Comment by smartekIT** [comment]
<pre>
 info
Current Core: Mar 25 2022, 1826524908
Agent Time: 2022-08-31 16:05:41.645+10:00.
</pre>

**Comment by si458** [comment]
> info
> Current Core: Mar 25 2022, 1826524908
> Agent Time: 2022-08-31 16:05:41.645+10:00.

the remote device isnt being updated for some reason?
![gif](https://media3.giphy.com/media/DUtVdGeIU8lmo/giphy.gif)

**Comment by smartekIT** [comment]
Ok i figured it out why the agent on the server not updating:
on the server under <pre>meshcentral-data</pre> folder, we have 2 folders: 
- signedagents (containing the agents signed by built-in meshcentral authinticode)
- agents (containing the agents I signed locally with my EV code sign cert)
in the past we requested from @Ylianst to keep the folder <pre>meshcentral-data/agents</pre> and give it the priority over <pre>meshcentral-data/signedagents</pre> folder so we still can use our EV code sign cert to sign the agents. which all good.

now the normal behaviour is, Authinticode edit and sign agents in <pre>node_modules/meshcentral/agents</pre>, then copy them to <pre>meshcentra-data/signedagents</pre> folder. (this is important as Authinticode will change icons as we required)

then we: 
- download files in <pre>meshcentra-data/signedagents</pre>
- we sign these files locally, then upload them to <pre>meshcentral-data/agents</pre>
- and as @Ylianst made <pre>meshcentral-data/agents</pre> on a higher priority, so our agents signed with our EV code cert will be used.

the above structure is perfect and working very well, until i think June. I see that Authiticode instead of taking the new updated agents files from <pre>node_modules/meshcentral/agents</pre> and sign them with copy to <pre>meshcentra-data/signedagents</pre>

it's actually looking at <pre>meshcentral-data/agents</pre> and take the agents files from there and sign them then copy to <pre>meshcentra-data/signedagents</pre> which is absolutely wrong.

so what I just did now, is deleted my files at <pre>meshcentra-data/signedagents</pre>
also delete my files at <pre>meshcentra-data/agents</pre>

rebooted the server, so authinticode couldn't see any files in <pre>meshcentra-data/agents</pre>, so it was forced to take the correct updated agents from <pre>node_modules/meshcentral/agents</pre>

**Comment by smartekIT** [comment]
so my suggestion @Ylianst is you keep the server looking at <pre>meshcentral-data/agents</pre> with a higher priority for download and pushing updates, But making sure when Authinticode wants to sign files after update, it takes them from  <pre>node_modules/meshcentral/agents</pre> ignoring if there is files in <pre>meshcentral-data/agents</pre>

then we can manually (locally) take the files from <pre>meshcentral-data/signedagents</pre> sign them and re-upload them to <pre>meshcentral-data/agents</pre>

That will fix it all I guess.

**Comment by si458** [comment]
@smartekIT glad it sorted now, thinking about it,
i did these steps...
1. update server and reboot
2. copied the windows agents from node_modules/meshcentral/agents/ locally
3. signed it with ezsignit
4. copied them back to meshcentral-data/agents/
5. restarted server again
6. updated agents manually as i have it disabled for autoupdate of agents

**Comment by smartekIT** [comment]
thanks, as said the problem was where meshcentral built-in authinticode takes the agents from to sign them! it should take them from the original place node_modules/meshcentral/agents/
now it's doing it wrong, as when there is files in meshcentral-data/agents/ it takes them from there!!

also for you when you want to sign your agents locally (because you have EV code cert and must use a USB key). you shouldn't take the files from node_modules/meshcentral/agents/
you should take them from meshcentral-data/signedagents/
because the built-in meshcentral Authinticode didn't just add a cert to the agents, but it did modify the icons files as you configured in config.json then it dropped the final and signed agents to meshcentral-data/signedagents/

unless you have no icon and info to be changed on the agent then you can copy them to local laptop from  modules/meshcentral/agents/

hope that make sense. I know it's hard t explain it, but I did my best and hope you guys got it.

Thanks heaps, we should wait on @Ylianst answer on fixing the authinticode tool bug.

**Comment by si458** [comment]
> also for you when you want to sign your agents locally (because you have EV code cert and must use a USB key). you shouldn't take the files from node_modules/meshcentral/agents/

We only have a basic certificate not an ev one

Also I have to sign it locally because stupid Norton keeps detecting the authenticode signed ones as viruses no matter if we say its not a virus

> you should take them from meshcentral-data/signedagents/
> because the built-in meshcentral Authinticode didn't just add a cert to the agents, but it did modify the icons files as you configured in config.json then it dropped the final and signed agents to meshcentral-data/signedagents/
> 

Also I had forgotten about the custom icon change so maybe try using authenticode again and see if Norton still likes it or not?

**Comment by smartekIT** [comment]
for symantec product, in the past i had to submit my files to them so they can whitelisted it which they did. actually symantec are pretty quick answering you back. here is the link to submit your files:
https://symsubmit.symantec.com/false_positive?lang=en

and here is the link to submit to microsoft as well which i normally do;
https://www.microsoft.com/en-us/wdsi/filesubmission

**Comment by smartekIT** [comment]
also a hit for you. if you have a standard code sign cert, you don't need to download the files to sign them locally. there is a tool for linux called Jsign which will sign them for you with all the options needed if you don't want to use the built-in authinticode tool.
i used to use both. in the past I made a script to use jsign and sign my agents on the server even with URL lock option using my standard code sign cert.
also i used the built-in authinticode tool to sign with lock option and also using my standard code cert and all happening on the linux server without having to download anything to sign locally.
our Hero @Ylianst has listened to us in the past and added an option to the to the config.json where we can tell the server to run a script after each update. (which comes handy to tell it to run our jsign script).
<pre>
    "_runOnServerStarted": "c:\\tmp\\mcstart.bat",
    "runOnServerUpdated": "put jsign script path here",
    "_runOnServerError": "c:\\tmp\\mcerror.bat",
</pre>

**Comment by smartekIT** [comment]
let me know if you need help installing and configuring JSign including make the script to sign with your standard code cert on linux box where meshcentral installed.

**Comment by smartekIT** [comment]
@krayon007 I just tried the taskbar option, and yes it works but for Windows TaskBar!!, I think you misunderstood me. what we were talking about here previously is the Privacybar not the windows taskbar.

**Comment by silversword411** [comment]
> That's an old version, you need to issue `agentupdate`

Can you post what an updated agents output is? I'm seeing March here too and `agentupdate` doesn't make any changes.

**Comment by si458** [comment]
> > That's an old version, you need to issue `agentupdate`
> 
> Can you post what an updated agents output is? I'm seeing March here too and `agentupdate` doesn't make any changes.

LINUX
```
> info
Current Core: Aug 29 2022, 294730201
Agent Time: 2022-08-31 12:06:36.953+01:00.
User Rights: 0xffffffff.
Platform: linux.
Capabilities: 15.
Server URL: wss://meshcentral.mefoo.com:443/agent.ashx.
OS: Ubuntu 22.04.1 LTS.
Modules: amt-apfclient, amt-lme, amt-manage, amt-mei, linux-dhcp, monitor-border, smbios, sysinfo, util-agentlog, wifi-scanner-windows, wifi-scanner.
Server Connection: true, State: 1.
X11 support: true.
> versions
{
  "openssl": "1.1.1q",
  "duktape": "v2.6.0",
  "commitDate": "2022-08-25T07:17:18.000Z",
  "commitHash": "f221183413f39aba155c75dbb85ff72777fc8244",
  "compileTime": "10:49:00, Aug 29 2022"
}
```

WINDOWS
```
> info
Current Core: Aug 25 2022, 1360394467
Agent Time: 2022-08-31 12:07:06.863+01:00.
User Rights: 0xffffffff.
Platform: win32.
Capabilities: 15.
Server URL: wss://meshcentral.mefoo.com:443/agent.ashx.
OS: Microsoft Windows 10 Pro - 21H2/19044.
Modules: amt-apfclient, amt-lme, amt-manage, amt-mei, monitor-border, smbios, sysinfo, util-agentlog, wifi-scanner-windows, wifi-scanner, win-console, win-info, win-securitycenter, win-terminal, win-virtual-terminal, win-volumes.
Server Connection: true, State: 1.
> versions
{
  "openssl": "1.1.1q",
  "duktape": "v2.6.0",
  "commitDate": "2022-08-25T07:17:18.000Z",
  "commitHash": "f221183413f39aba155c75dbb85ff72777fc8244",
  "compileTime": "10:55:09, Aug 25 2022"
}
```

**Comment by silversword411** [comment]
So `Current Core` is basically your mesh version...in build time notation :)

**Comment by dinger1986** [comment]
@si458 can close
